diff --git a/doc/main.h b/doc/main.h
--- a/doc/main.h
+++ b/doc/main.h
@@ -51,6 +51,7 @@
  *     - nix-vector-routing
  *     - ns3tcp
  *     - ns3wifi
+ *     - ofswitch13
  *     - olsr
  *     - openflow
  *     - point-to-point
diff --git a/doc/models/Makefile b/doc/models/Makefile
--- a/doc/models/Makefile
+++ b/doc/models/Makefile
@@ -67,6 +67,10 @@
 	$(SRC)/internet/doc/udp.rst \
 	$(SRC)/internet-apps/doc/internet-apps.rst \
 	$(SRC)/mobility/doc/mobility.rst \
+	$(SRC)/ofswitch13/doc/source/ofswitch13.rst \
+	$(SRC)/ofswitch13/doc/source/ofswitch13-description.rst \
+	$(SRC)/ofswitch13/doc/source/ofswitch13-usage.rst \
+	$(SRC)/ofswitch13/doc/source/ofswitch13-validation.rst \
 	$(SRC)/olsr/doc/olsr.rst \
 	$(SRC)/openflow/doc/openflow-switch.rst \
 	$(SRC)/point-to-point/doc/point-to-point.rst \
@@ -285,6 +289,12 @@
 	$(SRC)/netanim/doc/figures/RoutingTables.pdf \
 	$(SRC)/netanim/doc/figures/PacketTimeline.png \
 	$(SRC)/netanim/doc/figures/PacketTimeline.pdf \
+	$(SRC)/ofswitch13/doc/source/figures/ofswitch13-controller.eps \
+	$(SRC)/ofswitch13/doc/source/figures/ofswitch13-library.eps \
+	$(SRC)/ofswitch13/doc/source/figures/ofswitch13-module.eps \
+	$(SRC)/ofswitch13/doc/source/figures/ofswitch13-qos-topology.eps \
+	$(SRC)/ofswitch13/doc/source/figures/ofswitch13-queue.eps \
+	$(SRC)/ofswitch13/doc/source/figures/ofswitch13-switch.eps \
 	$(SRC)/spectrum/doc/spectrum-channel-phy-interface.png \
 	$(SRC)/spectrum/doc/spectrum-channel-phy-interface.pdf \
 	$(SRC)/spectrum/doc/spectrum-analyzer-example.eps \
@@ -382,6 +392,12 @@
 	$(FIGURES)/lr-wpan-primitives.eps \
 	$(FIGURES)/lr-wpan-data-example.eps \
 	$(FIGURES)/lr-wpan-arch.eps \
+	$(FIGURES)/ofswitch13-controller.eps \
+	$(FIGURES)/ofswitch13-library.eps \
+	$(FIGURES)/ofswitch13-module.eps \
+	$(FIGURES)/ofswitch13-qos-topology.eps \
+	$(FIGURES)/ofswitch13-queue.eps \
+	$(FIGURES)/ofswitch13-switch.eps \
 	$(FIGURES)/802-15-4-ber.eps \
 	$(FIGURES)/802-15-4-per-sens.eps \
 	$(FIGURES)/802-15-4-psr-distance.eps \
@@ -449,6 +465,12 @@
 $(FIGURES)/spectrum-tv-rand-geo-points.pdf_width = 8cm
 $(FIGURES)/lr-wpan-primitives.pdf_width = 3in
 $(FIGURES)/lr-wpan-arch.pdf_width = 2in
+$(FIGURES)/ofswitch13-controller.pdf_width = 10cm
+$(FIGURES)/ofswitch13-library.pdf_width = 14cm
+$(FIGURES)/ofswitch13-module.pdf_width = 14cm
+$(FIGURES)/ofswitch13-qos-topology.pdf_width = 12cm
+$(FIGURES)/ofswitch13-queue.pdf_width = 10cm
+$(FIGURES)/ofswitch13-switch.pdf_width = 10cm
 $(FIGURES)/clear-channel.pdf_width = 12cm
 $(FIGURES)/nist-frame-success-rate.pdf_width = 12cm
 $(FIGURES)/nist-frame-success-rate-n.pdf_width = 12cm
diff --git a/doc/models/source/index.rst b/doc/models/source/index.rst
--- a/doc/models/source/index.rst
+++ b/doc/models/source/index.rst
@@ -40,6 +40,7 @@
    mobility
    network
    nix-vector-routing
+   ofswitch13
    olsr
    openflow-switch
    point-to-point
diff --git a/doc/ns3_html_theme/get_version.sh b/doc/ns3_html_theme/get_version.sh
--- a/doc/ns3_html_theme/get_version.sh
+++ b/doc/ns3_html_theme/get_version.sh
@@ -191,7 +191,7 @@
     echo "// ns3_version.js:  automatically generated"       >  $outf
     echo "//  private urls"                                  >> $outf
     echo "var ns3_host = \"file://$PWD/\";"                  >  $outf
-    echo "var ns3_version = \"$repo @ $version$dirty\";"     >> $outf
+    echo "var ns3_version = \"ns-3.29 + OFSwitch13 3.2.2\";" >> $outf
     echo "var ns3_release = \"doc/\";"                       >> $outf
     echo "var ns3_local = \"build/\";"                       >> $outf
     echo "var ns3_doxy  = \"html/\";"                        >> $outf
